### PR TITLE
Fix revocations logic

### PR DESCRIPTION
**Pull request by justprosh**
Previously revocation works only in a direct way, just revoke the previous trust A -> B. 
Now it revokes all chains which contain a path from A to B, for example, A -> C -> D -> B
Also changed `Revoke` structure naming to plural `Revocation`

### CODE REVIEW

File: service/src/storage_impl.rs [hunk 1]
```diff
@@ -85,6 +85,32 @@ impl SQLiteStorage {
...
+            .cursor();
+
+        cursor.bind(&[
+            Value::String(format!("{}", issued_for)),
+            Value::Integer(relation_type),
+        ])?;
+        let mut relations: Vec<TrustRelation> = vec![];
+
+        while let Some(row) = cursor.next()? {
+            relations.push(parse_relation(row)?);
```
Comment (folex) [inline service/src/storage_impl.rs:109]: Are you sure you want to exit on any incorrect relation? Not very fault-tolerant

File: service/src/storage_impl.rs [hunk 2]
```diff
@@ -189,25 +215,36 @@ impl Storage for SQLiteStorage {
...
-        }
+    fn get_authorizations(&self, issued_for: &PublicKeyHashable) -> Result<Vec<Auth>, Self::Error> {
+        Ok(self
+            .get_relations(issued_for, AUTH_TYPE)?
+            .into_iter()
+            .fold(vec![], |mut acc, r| {
+                match r {
+                    TrustRelation::Auth(a) => acc.push(a),
+                    _ => (),
+                };
```
Comment (folex) [inline service/src/storage_impl.rs:226]: Maybe better use `if let`? 

Also, please run clippy

File: service/src/storage_impl.rs [hunk 3]
```diff
@@ -189,25 +215,36 @@ impl Storage for SQLiteStorage {
...
+        issued_for: &PublicKeyHashable,
+    ) -> Result<Vec<Revocation>, Self::Error> {
+        Ok(self
+            .get_relations(issued_for, REVOCATION_TYPE)?
+            .into_iter()
+            .fold(vec![], |mut acc, r| {
+                match r {
+                    TrustRelation::Revocation(revocation) => acc.push(revocation),
+                    _ => (),
+                };
```
Comment (folex) [inline service/src/storage_impl.rs:244]: ditto

File: service/src/tests.rs [hunk 4]
```diff
@@ -469,6 +469,54 @@ mod service_tests {
         assert_eq!(weight, 0u32);
     }
 
+    #[test]
+    fn revoke_test_updated() {
```
Comment (folex) [inline service/src/tests.rs:473]: what does 'updated' mean?

File: src/trust_graph.rs [hunk 5]
```diff
@@ -244,13 +244,42 @@ where
         pk: &PK,
         roots: HashSet<&PK>,
     ) -> Result<Vec<Vec<Auth>>, TrustGraphError> {
+        #[derive(Clone)]
+        struct Chain {
+            auths: Vec<Auth>,
```
Comment (folex) [inline src/trust_graph.rs:249]: > .expect("`cur_chain` always has at least one element");

maybe use non-empty vec? Like this one https://lib.rs/crates/nonempty

File: src/trust_graph.rs [hunk 6]
```diff
@@ -244,13 +244,42 @@ where
...
+                !self.revoked_by.contains(pk.as_ref())
+                    && !self.auths.iter().any(|a| a.trust.issued_for.eq(pk))
+            }
+
+            fn add_revocations(&mut self, revocations: Vec<Revocation>) {
+                revocations.into_iter().for_each(move |r| {
+                    self.revoked_by.insert(r.revoked_by.clone().into());
+                });
+            }
+        }
```
[the remark this answers is not in the document]
Comment (justprosh) [reply src/trust_graph.rs:272]: Maybe but it is only used in one separate method

File: service/src/storage_impl.rs (hunk 1 above)
Comment (justprosh) [reply service/src/storage_impl.rs:109]: If we can't parse the current relation something terrible already happened in serialization previously.
What do you suggest? Just skip incorrect?

File: service/src/tests.rs (hunk 4 above)
Comment (justprosh) [reply service/src/tests.rs:473]: This test is about updated revocation behaviour

File: src/trust_graph.rs [hunk 7]
```diff
@@ -270,13 +300,14 @@ where
                 .get_authorizations(&last.issued_by.clone().into())?;
 
             for auth in auths {
-                // if there is auth, that we not visited in the current chain, copy chain and append this auth
-                if !cur_chain
-                    .iter()
-                    .any(|a| a.trust.issued_for == auth.issued_by)
-                {
+                // if there is auth, that we not visited in the current chain and no revocations to any chain member --  copy chain and append this auth
```
Comment (folex) [inline src/trust_graph.rs:303]: why? let's document the motivation and semantics, not the literal meaning of the code

File: src/trust_graph.rs [hunk 8]
```diff
@@ -334,8 +365,8 @@ where
     }
 
     /// Mark public key as revoked.
-    pub fn revoke(&mut self, revoke: Revoke) -> Result<(), TrustGraphError> {
-        Revoke::verify(&revoke)?;
+    pub fn revoke(&mut self, revoke: Revocation) -> Result<(), TrustGraphError> {
+        Revocation::verify(&revoke)?;
```
Comment (folex) [inline src/trust_graph.rs:369]:
```suggestion
    pub fn revoke(&mut self, revocation: Revocation) -> Result<(), TrustGraphError> {
        Revocation::verify(&revocation)?;
```

File: src/trust_graph_storage.rs [hunk 9]
```diff
@@ -17,12 +17,14 @@ pub trait Storage {
...
     fn get_authorizations(&self, pk: &PK) -> Result<Vec<Auth>, Self::Error>;
+    fn get_revocations(&self, pk: &PK) -> Result<Vec<Revocation>, Self::Error>;
+
     fn insert(&mut self, node: TrustRelation) -> Result<(), Self::Error>;
 
     fn get_root_weight_factor(&self, pk: &PK) -> Result<Option<WeightFactor>, Self::Error>;
     fn add_root_weight_factor(&mut self, pk: PK, weight: WeightFactor) -> Result<(), Self::Error>;
     fn root_keys(&self) -> Result<Vec<PK>, Self::Error>;
-    fn revoke(&mut self, revoke: Revoke) -> Result<(), Self::Error>;
+    fn revoke(&mut self, revoke: Revocation) -> Result<(), Self::Error>;
```
Comment (folex) [inline src/trust_graph_storage.rs:27]:
```suggestion
    fn revoke(&mut self, revocation: Revocation) -> Result<(), Self::Error>;
```

File: src/trust_relation.rs [hunk 10]
```diff
@@ -33,44 +33,44 @@ pub struct Auth {
...
+    Revocation(Revocation),
 }
 
 impl TrustRelation {
     /// Returns timestamp of when this relation was created
     pub fn issued_at(&self) -> Duration {
         match self {
             TrustRelation::Auth(auth) => auth.trust.issued_at,
-            TrustRelation::Revoke(revoke) => revoke.revoked_at,
+            TrustRelation::Revocation(revoke) => revoke.revoked_at,
```
Comment (folex) [inline src/trust_relation.rs:44]: revoke => revocation?

Comment (folex) [reply src/trust_relation.rs:44]: or just `r`

File: service/src/storage_impl.rs (hunk 1 above)
Comment (folex) [reply service/src/storage_impl.rs:109]: I think skipping and logging is more fault-tolerant, so I'd go this way. But I don't know the overall picture, so it's up to you.

File: service/src/tests.rs (hunk 4 above)
Comment (folex) [reply service/src/tests.rs:473]: So the "updated" part means something only to 1-2 people. Let's name this test more meaningfully.

File: src/trust_graph.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (folex) [reply src/trust_graph.rs:272]: And it always will be used in a single method if you don't separate it :) If you do, there's a chance you'll find some other place to apply it.

And anyway:
- too much indentation
- already too much lines for a single file

File: src/trust_graph.rs (hunk 5 above)
Comment (folex) [reply src/trust_graph.rs:249]: I'm talking about `Chain::auths` field. It's not a `Queue`, it's `Vec`
image

Comment (justprosh) [reply src/trust_graph.rs:249]: Sorry, u r right

**Review by folex — APPROVED** [review]

**Pull request merged**